libraries/libfc/include/fc/io/random_access_file.hpp

```diff
@@ -0,0 +1,465 @@
+         return read(&c, 1);
+      }
+
+   private:
+      boost::beast::multi_buffer           buffer;
+      impl::random_access_file_context_ptr ctx;
+      ssize_t                              next_pos;
+   };
+
+   class write_datastream {
+      friend class random_access_file;
+      constexpr static unsigned buffer_flush_threshold = 256*1024;
+
+      write_datastream(const write_datastream&) = delete;
+      write_datastream& operator=(const write_datastream&) = delete;
+
+      write_datastream(impl::random_access_file_context_ptr& ctx, const ssize_t start_pos) : ctx(ctx), next_pos(start_pos) {
+         buffer.reserve(buffer_flush_threshold + 4*1024);
+      }
+
+      void do_write() {
+         ctx->write_to(buffer.cdata(), next_pos);
+         if(next_pos != impl::append_t)
+            next_pos += buffer.size();
+         buffer.clear();
+      }
+
+      void do_write_if_buffer_large() {
+         if(buffer.size() > buffer_flush_threshold)
+            do_write();
+      }
+
+   public:
+      write_datastream(write_datastream&& o) = default;
+
+      bool write(const char* d, size_t s) {
+         boost::asio::buffer_copy(buffer.prepare(s), boost::asio::const_buffer(d, s));
+         buffer.commit(s);
+
+         do_write_if_buffer_large();
+         return true;
+      }
+
+      inline bool put(char c) {
+         return write(&c, sizeof(char));
+      }
+      ~write_datastream() {
+         if(ctx)
+            do_write();
+      }
+
+   private:
+      boost::beast::flat_buffer                buffer;
+      impl::random_access_file_context_ptr     ctx;
+      ssize_t                                  next_pos;
+   };
+
+   class device {
+   public:
+      friend class random_access_file;
+      typedef char char_type;
+      typedef boost::iostreams::seekable_device_tag category;
+
+      explicit device(impl::random_access_file_context_ptr& ctx) : ctx(ctx) {}
+
+      //"To be usable with the streams and stream buffers provided by the Boost Iostreams library, Devices must model Blocking."
+      //"A Device is Blocking if a read request never produces fewer characters than requested except at end-of-stream"
+      std::streamsize read(char* s, std::streamsize n) {
+         ssize_t total_red = 0;
+         while(n - total_red) {
+            ssize_t red = ctx->read_from(boost::asio::buffer(s+total_red, n-total_red), pos);
+            if(red == 0)
+               break;
+            pos += red;
+            total_red += red;
+         }
+         return total_red ?: -1;
+      }
+
+      std::streamsize write(const char* s, std::streamsize n) {
+         ctx->write_to(boost::asio::const_buffer(s, n), pos);
+         pos += n;
+         return n;
+      }
+
+      std::streampos seek(boost::iostreams::stream_offset off, std::ios_base::seekdir way) {
+         if(way == std::ios_base::beg)
+            pos = off;
+         else if(way == std::ios_base::end)
+            pos = ctx->size() + off;
+         else
```

Comment:
No I don't think so -- writing past the end of a file leaving a gap is well defined on any modern POSIX-like OS I believe. There is even a unit test for it with a datastream,
https://github.com/AntelopeIO/spring/blob/a29f3f96aa9bfebc3a33eb4c3578bab74bb08527/libraries/libfc/test/io/test_random_access_file.cpp#L25-L28
(there is an 8 byte gap between the write on line 25 and line 27 that is checked a few lines further down)

Perhaps there should be a test for the device as well instead of just the datastream, though the semantics are all the same (ultimately calling `pwrite()`).

I admit I'm not sure if this is defined behavior on Win32[^1]. fwiw the test above _did_ pass on Win32 empirically.

[^1]: `SetEndOfFile()` states that extending a file past the end leaves the new contents undefined, but I don't see anything in `WriteFileEx()` that states what happens when writing past the end.